pkg/master/master.go

```diff
@@ -544,7 +554,34 @@ func (m *Master) init(c *Config) {
 	}
 
 	// TODO: Attempt clean shutdown?
-	m.masterServices.Start()
+	if m.enableCoreControllers {
+		m.NewBootstrapController().Start()
+	}
+}
+
+// NewBootstrapController returns a controller for watching the core capabilities of the master.
+func (m *Master) NewBootstrapController() *Controller {
```

Comment:
If the master was inverted (sets of simple overlapping functionality brought together by a common, simple initializer) you wouldn't get the warning.  It's really just that the master is designed as the blackhole that sucks in all other function, which means there is no opportunity to consume it flexibly.  I don't want to make the master a leaky box, I really would just prefer it be composed better (for pure code cleanliness) which also would have the benefit that an opinionated customizer could do a bit more with it without having to carry patches.  I think the Hyperkube consumption use case is almost exactly the openshift use case, with just a slight variation on arguments.

----- Original Message -----
> > @@ -544,7 +554,34 @@ func (m *Master) init(c *Config) {
> >  	}
> >  
> >  	// TODO: Attempt clean shutdown?
> > -	m.masterServices.Start()
> > +	if m.enableCoreControllers {
> > +		m.NewBootstrapController().Start()
> > +	}
> > +}
> > +
> > +// NewBootstrapController returns a controller for watching the core
> > capabilities of the master.
> > +func (m *Master) NewBootstrapController() *Controller {
> 
> I see.  I am sympathetic, but I want to be really careful about leaving
> debris around the kube codebase to accomodate out-of-tree stuff.  This is
> the sort of thing that is easily broken when some static analysis tool says
> "hey this could be private" and then we change it.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7704/files#r29891728